### DISCUSSION TITLE
Highlighting spaces and tabs

### DISCUSSION THREAD

**Discussion by thaliaarchi** [github-linguist/linguist#5519] [Q&A] [answered]
I'd like to add Linguist support for the Whitespace programming language. It should meet the [200-repo](https://github.com/wspace/corpus) requirement and there is precedent for including other esolangs in Linguist (e.g. Brainfuck, Befunge, LOLCODE).

The only characters in Whitespace are space, tab, and line feed, but since they are non-printable, there is no text to highlight. There are two existing TextMate grammars for Whitespace: [veniversum/language-whitespace](https://github.com/veniversum/language-whitespace) and [bvisness/whitespace](https://github.com/bvisness/whitespace/tree/master/misc/sublime). @veniversum appears to solve this by replacing space, tab, and line feed with `s`, `t`, and `¬`, respectively, then coloring those. @bvisness creates a theme that colors the background color of those tokens. Is it possible to set background colors for GitHub syntax highlighting?

**Comment by bvisness** [comment]
It's one thing to support syntax highlighting for esolangs, but another thing to build new features specifically for esolangs. That seems like a clear waste of time. 

Besides, the only reason I did anything with Whitespace was because I was in college, it was on the Wikipedia page for esolangs, and it sounded funny. I never even got it running. Nothing I have done is a worthwhile precedent for serious work. 

And besides, isn't whitespace funnier when you can't see the contents of the file? 😛

**Reply by thaliaarchi** [reply]
Well as it turns out, I've seen several schools that use Whitespace in their compilers courses, so it's at least useful for pedagogical reasons.

**Comment by Nixinova** [comment, marked as answer]
Here's the CSS classes that have a background color on GitHub:

- `.pl-ii {color:#fafbfc;background-color:#b31d28}` - `#b31d28`
- `.pl-c2 {color:#fafbfc;background-color:#d73a49}` - `#d73a49`
- `.pl-md {color:#b31d28;background-color:#ffeef0}` - `#ffeef0`
- `.pl-mi1{color:#22863a;background-color:#f0fff4}` - `#f0fff4`
- `.pl-mc {color:#e36209;background-color:#ffebda}` - `#ffebda`
- `.pl-mi2{color:#f6f8fa;background-color:#005cc5}` - `#005cc5`

So you (or I) could create a grammar that uses `.pl-ii` (token `invalid.illegal`) and `.pl-mi2` (not sure what token) and display well on GitHub.

**Reply by thaliaarchi** [reply]
So if these were to be used, the grammar tokens couldn't be `keyword.control.space` and `keyword.control.tab`, but rather `invalid.illegal` and <token corresponding to `.pl-mi2`>?

If I get the go-ahead to add Whitespace, I'll create a repo for the grammar.

**Reply by Nixinova** [reply]
Ah, found a token&rarr;class map: https://github.com/aschoettler/misc-stuff/blob/349b824d321f15286f6af3478bc52a1eeaa10770/github-light.map.json

`mi2` is 'markup.ignored'

**Reply by Nixinova** [reply]
Here you go, a simple grammar for highlighting WhiteSpace files:

- [Lightshow](https://github-lightshow.herokuapp.com/?utf8=%E2%9C%93&scope=from-text&grammar_format=auto&grammar_url=&grammar_text=name%3A+WhiteSpace+Esolang%0D%0AscopeName%3A+source.ws%0D%0Apatterns%3A%0D%0A-+name%3A+markup.ignored.character.space.ws%0D%0A++match%3A+%27+%27%0D%0A-+name%3A+invalid.illegal.character.tab.ws%0D%0A++match%3A+%27%5Ct%27%0D%0A-+name%3A+carriage-return.ws%0D%0A++match%3A+%27%5Cr%27&code_source=from-text&code_url=&code=++%09+++%09%0D%0A++%09+++++%09)
![image](https://user-images.githubusercontent.com/42429413/129466008-75ed9d65-afe6-4828-b8ee-879f5a1c3c72.png)

**Reply by thaliaarchi** [reply]
I checked each of the other tokens. For, `carriage-return`, it renders a `^M` for each instance. The best case for this grammar would be to have a middle dot `· ` rendered for space and an arrow `→` for tab.

```
.pl-ii  invalid.illegal
.pl-c2  carriage-return ^M
.pl-md  markup.deleted
.pl-mi1 markup.inserted
.pl-mc  markup.changed
.pl-mi2 markup.ignored
```

**Reply by Nixinova** [reply]
[This](https://github-lightshow.herokuapp.com/assets/application-fe1f67824c97362650f12ea36c9746ee2006db70301ff0f774ca6d4a2b0e4d85.css) is all GitHub gives you. The only class that adds a character is `carriage-return`, '^M'.

**Reply by thaliaarchi** [reply]
It's probably too much to ask for two more classes then, right?

**Reply by Alhadis** [reply]
[Bookmark this](https://github.com/Alhadis/language-etc/blob/master/samples/lists/scope-previews.nanorc), and bookmark it well. This is an exhaustive list not just of the TextMate-scope/CSS-class mappings, but also of their current styling (which IS subject to wanton, random changes at the whims of GitHub's designer team, JSYK).

> So if these were to be used, the grammar tokens couldn't be keyword.control.space and keyword.control.tab, but rather invalid.illegal and <token corresponding to .pl-mi2>?

There's actually a trick you can use if you're willing to complicate your grammar a little. It exploits the fact that TextMate scope selectors are order dependent; i.e., unlike CSS selectors, a TextMate scope of `keyword.operator` *isn't* the same as a CSS class selector for `.operator.keyword` (although something like `keyword.operator.whatever` will work).

Hence, you can "hide" a token from GitHub's highlighter by prefixing it with a scope it doesn't recognise. For example:

~~~cson
tab:
	name: "markup.inserted.whitespace"
	match: "\t"
	captures:
		0: name: "_.keyword.control.tab.whitespace"

space:
	name: "markup.changed.space.whitespace"
	match: " "
	captures:
		0: name: "_.keyword.control.space.whitespace"
~~~

[Here's a preview](https://github-lightshow.herokuapp.com/?utf8=%E2%9C%93&scope=from-text&grammar_format=cson&grammar_url=&grammar_text=name%3A+%22Whitespace%22%0D%0AscopeName%3A+%22source.whitespace%22%0D%0AfileTypes%3A+%5B%22.ws%22%5D%0D%0Apatterns%3A+%5Binclude%3A+%22%23main%22%5D%0D%0A%0D%0Arepository%3A%0D%0A%09main%3A%0D%0A%09%09patterns%3A+%5B%0D%0A%09%09%09%7Binclude%3A+%22%23tab%22%7D%0D%0A%09%09%09%7Binclude%3A+%22%23space%22%7D%0D%0A%09%09%5D%0D%0A%0D%0A%09tab%3A%0D%0A%09%09name%3A+%22markup.ignored.whitespace%22%0D%0A%09%09match%3A+%22%5Ct%22%0D%0A%09%09captures%3A%0D%0A%09%09%090%3A+name%3A+%22_.keyword.control.tab.whitespace%22%0D%0A%0D%0A%09space%3A%0D%0A%09%09name%3A+%22invalid.illegal.space.whitespace%22%0D%0A%09%09match%3A+%22+%22%0D%0A%09%09captures%3A%0D%0A%09%09%090%3A+name%3A+%22_.keyword.control.space.whitespace%22&code_source=from-text&code_url=&code=%09+++%09%09) with an unabridged example.

<details><summary><b>Show grammar source</b></summary>

~~~cson
name: "Whitespace"
scopeName: "source.whitespace"
fileTypes: [".ws"]
patterns: [include: "#main"]

repository:
	main:
		patterns: [
			{include: "#tab"}
			{include: "#space"}
		]

	tab:
		name: "markup.ignored.whitespace"
		match: "\t"
		captures:
			0: name: "_.keyword.control.tab.whitespace"

	space:
		name: "invalid.illegal.space.whitespace"
		match: " "
		captures:
			0: name: "_.keyword.control.space.whitespace"
~~~

</details>

**Comment by lildude** [comment]
> I'd like to add Linguist support for the Whitespace programming language. It should meet the 200-repo requirement and there is precedent for including other esolangs in Linguist (e.g. Brainfuck, Befunge, LOLCODE).

Keep in mind that we now have minimum usage requirements which may not have been in place when other esolangs were added. Given Whitespace has nothing to differentiate it content-wise you're only going to be able to rely on the extension or specific filename to prove usage.

**Reply by thaliaarchi** [reply]
There actually are some distinguishing features of Whitespace syntax: most lines in Whitespace have tabs after spaces, whereas tabs in other languages almost exclusively appear at the beginning of a line, only rarely after spaces, such as in coding styles that uses tabs for aligning tables. There's also very low prevalence of symbols, unless the program is a polyglot, in which case, it would make more sense to highlight the other language anyways.

**Reply by lildude** [reply]
Sorry, I should have made clear that you won't be able to demonstrate this with GitHub's Search.

**Reply by thaliaarchi** [reply]
I was afraid that would be the case. Is there a way to download all .ws files for local processing?

**Reply by lildude** [reply]
Unfortunately, not any more. The old internal method we used to use has been removed so you'd need to craft something yourself using the API. Warning: you're almost certain to hit rate limiting so this could take a very very long time.
